agent/api/task/task.go

```diff
@@ -1648,20 +1698,49 @@ func (task *Task) getASMAuthResource() ([]taskresource.TaskResource, bool) {
 	return res, ok
 }
 
-// PopulateSSMSecrets appends the container's env var map with ssm parameters
-func (task *Task) PopulateSSMSecrets(container *apicontainer.Container) *apierrors.DockerClientConfigError {
-	resource, ok := task.getSSMSecretsResource()
-	if !ok {
-		return &apierrors.DockerClientConfigError{"task secret data: unable to fetch SSM Secrets resource"}
+// getSSMSecretsResource retrieves ssmsecret resource from resource map
+func (task *Task) getSSMSecretsResource() ([]taskresource.TaskResource, bool) {
+	task.lock.RLock()
+	defer task.lock.RUnlock()
+
+	res, ok := task.ResourcesMapUnsafe[ssmsecret.ResourceName]
+	return res, ok
+}
+
+// PopulateSecretsAsEnv appends the container's env var map with secrets
+func (task *Task) PopulateSecretsAsEnv(container *apicontainer.Container) *apierrors.DockerClientConfigError {
+	var ssmRes *ssmsecret.SSMSecretResource
+	var asmRes *asmsecret.ASMSecretResource
+
+	if container.ShouldCreateWithSSMSecret() {
+		resource, ok := task.getSSMSecretsResource()
+		if !ok {
+			return &apierrors.DockerClientConfigError{"task secret data: unable to fetch SSM Secrets resource"}
+		}
+		ssmRes = resource[0].(*ssmsecret.SSMSecretResource)
+	}
+
+	if container.ShouldCreateWithASMSecret() {
+		resource, ok := task.getASMSecretsResource()
+		if !ok {
+			return &apierrors.DockerClientConfigError{"task secret data: unable to fetch ASM Secrets resource"}
+		}
+		asmRes = resource[0].(*asmsecret.ASMSecretResource)
 	}
 
-	ssmResource := resource[0].(*ssmsecret.SSMSecretResource)
 	envVars := make(map[string]string)
 
 	for _, secret := range container.Secrets {
-		if secret.Provider == apicontainer.SecretProviderSSM {
-			k := secret.GetSSMSecretResourceCacheKey()
-			if secretValue, ok := ssmResource.GetCachedSecretValue(k); ok {
+		if secret.Provider == apicontainer.SecretProviderSSM && secret.Type == apicontainer.SecretTypeEnv {
+			k := secret.GetSecretResourceCacheKey()
+			if secretValue, ok := ssmRes.GetCachedSecretValue(k); ok {
+				envVars[secret.Name] = secretValue
+			}
+		}
+
+		if secret.Provider == apicontainer.SecretProviderASM && secret.Type == apicontainer.SecretTypeEnv {
+			k := secret.GetSecretResourceCacheKey()
+			if secretValue, ok := asmRes.GetCachedSecretValue(k); ok {
```

Comment:
Does this mean ASM takes precedence over SSM secret if both have same keys? Shall we also add a debug statement if we are overriding already added values?

Reply:
It is not really necessary to add a debug message here. This situation will also apply to secret values overriding normal environment variables. We do not check duplicated names on either ECS backend or agent side.  Leave it to customers to verify by themselves.

Reply:
> This situation will also apply to secret values overriding normal environment variables.  

Don't we do this check in backend and throw error message? 

I agree we can leave it to customers to make sure the keys are unique. But we are making a decision here to override the SSM secret values with ASM secret values if the keys are same. We should somehow let customers know this through docs and adding logs here might be useful.